src/decoding/error.rs

```diff
@@ -1,52 +1,56 @@
-#[cfg(not(feature = "std"))]
-use alloc::{str::Utf8Error, string::FromUtf8Error};
-#[cfg(not(feature = "std"))]
-use core::num::ParseIntError;
-
 use alloc::{
     format,
-    string::{String, ToString},
+    str::Utf8Error,
+    string::{FromUtf8Error, String, ToString},
 };
-use core::fmt::{self, Display, Formatter};
+use core::{fmt::Display, num::ParseIntError};
 
 #[cfg(feature = "std")]
-use std::{error::Error as StdError, sync::Arc};
+use std::sync::Arc;
 
-use failure::Fail;
+use snafu::Snafu;
 
-use crate::state_tracker::StructureError;
+use crate::state_tracker;
 
-#[derive(Debug, Clone, Fail)]
+#[derive(Debug, Clone, Snafu)]
 pub struct Error {
-    #[fail(context)]
     context: Option<String>,
-    #[fail(cause)]
-    error: ErrorKind,
+    source: ErrorKind,
 }
 
-/// An enumeration of potential errors that appear during bencode deserialization.
-#[derive(Debug, Clone, Fail)]
+// An enumeration of potential errors that apepar during bencode deserialization.
+#[derive(Debug, Clone, Snafu)]
 pub enum ErrorKind {
     /// Error that occurs if the serialized structure contains invalid semantics.
     #[cfg(feature = "std")]
-    #[fail(display = "malformed content discovered: {}", _0)]
-    MalformedContent(Arc<failure::Error>),
+    #[snafu(display("malformed content discovered: {}", source))]
+    MalformedContent { source: Arc<dyn std::error::Error> },
+
     /// Error that occurs if the serialized structure contains invalid semantics.
     #[cfg(not(feature = "std"))]
-    #[fail(display = "malformed content discovered")]
+    #[snafu(display("malformed content discovered"))]
     MalformedContent,
+
     /// Error that occurs if the serialized structure is incomplete.
-    #[fail(display = "missing field: {}", _0)]
-    MissingField(String),
-    /// Error in the bencode structure (e.g. a missing field end separator).
-    #[fail(display = "bencode encoding corrupted ({})", _0)]
-    StructureError(#[fail(cause)] StructureError),
+    #[snafu(display("missing field: {}", field))]
+    MissingField { field: String },
+
+    /// Error in the bencode structure (e.g. a missing field and seperator).
+    #[snafu(display("bencode encoding corrupted ({})", source))]
```

Comment:
I think you can add `#[snafu(context(false), display(…))]`. This lets you convert from the source error type to the parent error type with `?`, without needing a from implementation. See the attributes guide for more info: https://docs.rs/snafu/0.6.9/snafu/guide/attributes/index.html

I think this also applies to every other error variant (if there are any) that only have `source` fields, and don't need to otherwise be disambiguated. (For example if they can occur in different contexts, and need to be mapped to different variants.)

Reply:
I did some quick tests and it seems like this doesn't work in case of our `Error` / `ErrorKind` combination, as the `context(false)` attribute only applies to the conversion from the inner source to the related `ErrorKind` enum and not to the outer `Error`.  As soon as I try to replace the `From` implementation the compiler isn't able anymore to convert, e.g. a `StructError` into the outer `Error` struct.

Reply:
Ahh, gotcha, I didn't think of that.